Spec update: Support multiple & cross-origin worklets

xyaoinum opened the pull request:
- selectURL() and run() will be exposed to the shared storage worklet object. When calling on the default scoped worklet (i.e. sharedStorage.worklet.selectURL()/run()), the behavior is equivalent to sharedStorage.selectURL()/run().
- Users can create new worklets via `const worklet = await sharedStorage.createWorklet(url, options)`. This can be used to start multiple and potentially cross-origin worklets from a single document.
- User settings error won't be exposed to the caller (i.e. will be treated as success) in the case of creating or using a cross-origin worklet. This is to prevent leaking user's settings for an arbitrary site.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/WICG/shared-storage/pull/131.html" title="Last updated on Apr 13, 2024, 2:47 AM UTC (9100129)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/shared-storage/131/ea33aba...9100129.html" title="Last updated on Apr 13, 2024, 2:47 AM UTC (9100129)">Diff</a>

xyaoinum commented:
@pythagoraskitty: PTAL. Thanks! (After it's looking okay from your end, I will also request review from a spec mentor.)

pythagoraskitty commented on spec.bs:
```diff
@@ -201,56 +201,227 @@ This document introduces a new storage API that is intentionally not partitioned
+      Here are the scenarios where this algorithm could be used:
+      - For <a href="#window-setter">Window setter/deleter methods</a>, |environment|'s [=environment settings object/origin=] should equal |origin|.
+      - For creating a worklet, and for initiating (from {{Window}}) and running (from {{SharedStorageWorkletGlobalScope}}) operations on a worklet, |environment| should be the worklet's creator window, and |origin| should be the [=SharedStorageWorklet/worklet origin=].
+      - For <a href="#ss-fetch-algo">fetch request</a>, |environment| should be the request's initiator window, and |origin| should be the request's URL's origin.
```
s/For <a href="#ss-fetch-algo">fetch request</a>/For a <a href="#ss-fetch-algo">fetch request</a>/

pythagoraskitty commented on spec.bs:
```diff
@@ -201,56 +201,227 @@ This document introduces a new storage API that is intentionally not partitioned
 
-    0. If {{Worklet}} has an associated boolean [=addModule initiated=], and the result of running [=check if addModule is allowed and update status=] on {{Worklet}} is false, return a [=promise rejected=] with a {{TypeError}}.
+    0. If |this| is of type {{SharedStorageWorklet}}, and the result of running [=check if addModule is allowed and update status=] given |this| and |moduleURL| is false:
+        1. If |this|'s [=SharedStorageWorklet/is cross-origin worklet=] is true, then reject <var ignore="">promise</var> with a {{TypeError}}. Else, resolve <var ignore="">promise</var>.
```
Yes, it looks like this is what you say below...

pythagoraskitty commented on spec.bs:
```diff
@@ -201,56 +201,227 @@ This document introduces a new storage API that is intentionally not partitioned
+
+    <div class="note">
+
+    On user preferences error, addModule() will be aborted at an early stage. However, the error will only be exposed to the caller for same-origin worklet (i.e. where the initiator document's origin is same-origin with the module script's origin). For cross-origin worklet, the error will be hidden. This is to prevent a caller from knowing which origins the user has disabled shared storage for vi ...
```
s/exposed to the caller for same-origin worklet/exposed to the caller for a same-origin worklet/

s/For cross-origin worklet/For a cross-origin worklet/

pythagoraskitty commented:
Suggest we move the {#reporting} and {#budgets} subsubsections of {#window} up to be subsections of the {#worklet} section instead. I have no strong preference about the ordering within that section, but you could probably just tack them on the end of it after {#scope-algo}.

pythagoraskitty commented:
I've looked at most, except need to give a closer read to the new versions of `selectURL()` and `run()` sometime when I am less tired.  :) 

Hopefully will get to in the next couple of days. Otherwise looks good % comments.

pythagoraskitty commented on spec.bs:
```diff
@@ -201,56 +201,227 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. Let |resultPromise| be a new [=promise=].
+    1. If [=this=]'s [=addModule initiated=] is false, then return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |window| be [=current global object=].
+    1. Assert that |window| is {{Window}}.
```
s/|window| is {{Window}}/|window| is a {{Window}}/

pythagoraskitty commented on spec.bs:
```diff
@@ -201,56 +201,227 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. Let |promise| be a new [=promise=].
+    1. If [=this=]'s [=addModule initiated=] is false, then return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |window| be [=current global object=].
+    1. Assert that |window| is {{Window}}.
```
s/|window| is {{Window}}/|window| is a {{Window}}/

pythagoraskitty commented on spec.bs:
```diff
@@ -201,56 +201,227 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. Let |context| be |window|'s [=Window/browsing context=].
+    1. If |context| is null, return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |document| be |context|'s [=active document=].
+    1. If the result of running [=Is feature enabled in document for origin?=] on "[=PermissionsPolicy/shared-storage=]", |document|, and [=this=]'s [=SharedStorageWorklet/worklet origin=] returns false, return a [=promise rejected=] with a {{TypeError}}.
```
Following this step, we probably also need the following? (I couldn't see where we are checking it otherwise.)

1. If the result of running [=Is feature enabled in document for origin?=] on "[=PermissionsPolicy/shared-storage-select-url=]", |document|, and [=this=]'s [=SharedStorageWorklet/worklet origin=] returns false, return a [=promise rejected=] with a {{TypeError}}.

xyaoinum replied:
I realized it should be "shared-storage-select-url" here, and the "shared-storage" policy is already checked in the "determine whether shared storage is allowed" algorithm.

However, I also realized this is no longer ideal with this new proposal. We only want to treat user preferences errors differently for same v.s. cross-origin worklets, whereas for context error (i.e. secure context, permissions policy), we do want to expose them as there's no privacy issue. For enrollment error, I'll categorize it as preferences error for now (as it's what Chrome does today), but it seems we can revisit this decision later.

Thus, I divided the [=determine whether shared storage is allowed=] to 2 algorithms: [=… allowed by context=] and [=… allowed by by enrollment and user preference=]. LMKWYT!

pythagoraskitty replied:
I think it's a good idea to split them up. :)

pythagoraskitty commented on spec.bs:
```diff
@@ -201,56 +201,227 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. [=Assert=] that [=this=]'s [=global scopes|list of global scopes=] [=list/contains=] a single {{SharedStorageWorkletGlobalScope}}.
+    1. If the result of running [=SharedStorageWorkletGlobalScope/check whether addModule is finished=] for [=this=]'s {{SharedStorageWorkletGlobalScope}} is false, return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |realm| be the [=current realm=].
+    1. Let |outsideSettings| be {{WindowSharedStorage/worklet}}'s [=relevant settings object=].
```
s/{{WindowSharedStorage/worklet}}'s [=relevant settings object=]/[=this=]'s [=relevant settings object=]/

pythagoraskitty commented on spec.bs:
```diff
@@ -201,56 +201,227 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. Let |agent| be the result of [=obtaining a worklet agent=] given |outsideSettings|.
+    1. Run the following steps in |agent|:
+        1. [=Queue a global task=] on the [=DOM manipulation task source=], given |realm|'s [=global object=], to resolve |promise| with undefined.
+        1. If {{WindowSharedStorage/worklet}}'s [=module map=] is not [=map/empty=]:
```
s/{{WindowSharedStorage/worklet}}'s [=module map=]/[=this=]'s [=module map=]/

pythagoraskitty commented on spec.bs:
```diff
@@ -201,56 +201,227 @@ This document introduces a new storage API that is intentionally not partitioned
+        1. Set |fencedFrameConfigStruct|'s [=/url=] to |urlList|[|resultIndex|].
+        1. Let |resultURLWithMetadata| be |urls|[|resultIndex|].
+        1. If |resultURLWithMetadata| has field "`reportingMetadata`", run [=register reporting metadata=] with |resultURLWithMetadata|["`reportingMetadata`"].
+        1. If |options|["`keepAlive`"] is false, run [=terminate a worklet global scope=] with the associated {{SharedStorageWorkletGlobalScope}}.
```
Maybe slightly better:

s/with the associated {{SharedStorageWorkletGlobalScope}}/with [=this=]'s {{SharedStorageWorkletGlobalScope}}/

pythagoraskitty commented on spec.bs:
```diff
@@ -201,56 +201,227 @@ This document introduces a new storage API that is intentionally not partitioned
+        1. If |options|["`keepAlive`"] is false, run [=terminate a worklet global scope=] with the associated {{SharedStorageWorkletGlobalScope}}.
+    1. [=Upon rejection=] of |indexPromise|, perform the following steps:
+        1. Set |fencedFrameConfigStruct|'s [=/url=] to |urlList|[[=default index=]].
+        1. If |options|["`keepAlive`"] is false, run [=terminate a worklet global scope=] with the associated {{SharedStorageWorkletGlobalScope}}.
```
Maybe slightly better:

s/with the associated {{SharedStorageWorkletGlobalScope}}/with [=this=]'s {{SharedStorageWorkletGlobalScope}}/

pythagoraskitty commented:
Finished checking. Looks good % comments. :)

xyaoinum commented:
Thanks Cammie. I did some more changes around dividing the permissions checks, as explained in the [comment above](https://github.com/WICG/shared-storage/pull/131#discussion_r1472158092). PTAL again!

@domfarolino : PTAL from the spec's perspective. Thanks!

xyaoinum commented:
@domfarolino : friendly ping! Also adding @wanderview  as an alternative reviewer, who may be able to cover the review this week when Dominic is OOO. Thanks! The idea is to make the spec good enough to unblock launch. We can track non-blocking issues separately if applicable.

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
-    1. If |worklet|'s [=addModule initiated=] is true, return false.
+    To <dfn>check if addModule is allowed and update status</dfn> given a {{SharedStorageWorklet}} |worklet| and an [=/URL=] |moduleURL|, run the following steps:
+    1. Let |workletOrigin| be |moduleURL|'s [=url/origin=].
+    1. If the result of running [=determine whether shared storage is allowed by context=] given [=this=]'s [=relevant settings object=] and |workletOrigin| is false, return "DisallowedDueToNonPreferenceError".
```
You can't really use `[=this=]` in a free-standing algorithm that's not a method. You'll need to pass in a global here.

Btw is the global that will be passed in (i.e., the `[=this=]` that you're already using) _always_ a `{{SharedStorageWorklet}}` object? It looks like it, given https://pr-preview.s3.amazonaws.com/WICG/shared-storage/131/ea33aba...6606cc0.html#worklet-monkey-patch.

If that's right, please `[=Assert=]` that up-front, just for clarity.

(This comment applies below to other errant uses of `[=this=]`)

xyaoinum replied:
Switched to [=current global object=]. This seems legitimate?

domfarolino replied:
@xyaoinum can you confirm whether this can be resolved or not?

xyaoinum replied:
Yes this can be resolved.

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
-    To <dfn>check if addModule is allowed and update status</dfn> for a {{SharedStorageWorklet}} |worklet|, run the following steps:
-    1. If the result of running [=determine whether shared storage is allowed=] on the [=relevant settings object=] of [=this=] is false, return false.
-    1. If |worklet|'s [=addModule initiated=] is true, return false.
+    To <dfn>check if addModule is allowed and update status</dfn> given a {{SharedStorageWorklet}} |worklet| and an [=/URL=] |moduleURL|, run the following steps:
```
super duper nit: "status" feels weird here. I might prefer "state" since we're updating multiple pieces of state here. No strong preference though, it's just that when I read "status" I found myself looking for one specific piece of status/enum or something, but I realized actually we're just making the whole global "ready" by setting it all up.

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
 
 Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=], initially empty, can contain at most one instance of its [=worklet global scope type=], the {{SharedStorageWorkletGlobalScope}}.
 
+  Note: The website that serves the module script should be aware of the implication of CORS: when the the module script's URL's origin is cross-origin with the worklet's creator window's origin, and by granting the module script resource via CORS, it will also grant the worklet's creation and subsequent operations on the worklet, under module script's URL's origin. The worklet's creator context  ...
```
Link to all kinds of things like URL/origin/cross-origin/ (can just use the `[=same origin=]` definition), "budget", and more please. There's lots of stuff here that a reader who really relies on this `Note:` could benefit from having direct links to.

domfarolino replied:
```suggestion
  Note: The website that serves the module script should be aware of the implication of CORS: when the the module script's URL's origin is cross-origin with the worklet's creator window's origin, and by granting the module script resource via CORS, it will also grant the worklet's creation and subsequent operations on the worklet, under module script's URL's origin. For example, the worklet's creator context could poison and use up the [=SharedStorageWorklet/worklet origin=]'s budget.
```

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+  ## Run Operation Methods on {{SharedStorageWorklet}} ## {#run-op-shared-storage-worklet}
+
+  <div algorithm>
+    To <dfn>get the select-url result index</dfn>, given {{SharedStorageWorklet}} |worklet|, {{DOMString}} |operationName|, [=/list=] |urlList|, and {{SharedStorageRunOperationMethodOptions}} |options|:
```
List of what? Often it's described in algorithm declarations.

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+        1. If [=this=]'s [=SharedStorageWorklet/is cross-origin worklet=] is false, return a [=promise rejected=] with a {{TypeError}}.
+    1. Else, set |fencedFrameConfigStruct| to a "fenced frame config struct" with a pending mapped "url uuid". <span class=todo>Add correct struct name as well as linking when Fenced Frame API updates their draft spec to include it.</span>
+
+    Issue: The "fenced frame config struct" and the following "obtain a {{FencedFrameConfig}} from a fenced frame config struct" algorithm have not yet been added to the draft [=Fenced Frame=] specification. Some form of them will be added, although their names are subject to bikeshedding.
```
Note that this issue breaks the algorithm step counter. Please indent this and confirm whether or not it fixes the step numbers. You can see that they're broken here: https://pr-preview.s3.amazonaws.com/WICG/shared-storage/131/ea33aba...6606cc0.html#worklet-monkey-patch:~:text=to%20include%20it.-,The%20%22fenced%20frame%20config%20struct%22%20and%20the%20following%20%22obtain,-a%20%20FencedFrameConfig%20%20from

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+
+    1. Let |promise| be a new [=promise=].
+    1. Let |window| be [=current global object=].
+    1. Assert that |window| is a {{Window}}.
```
```suggestion
    1. [=Assert=]: |window| is a {{Window}}.
```

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. Let |window| be [=current global object=].
+    1. Assert that |window| is a {{Window}}.
+    1. Let |context| be |window|'s [=Window/browsing context=].
+    1. If |context| is null, return a [=promise rejected=] with a {{TypeError}}.
```
Can you see what kind of error is most common to throw in this case? I feel like it is some Security-based DOMException (I see that in HTML a bit for this case), so I feel like TypeError might not be the best thing to throw here...

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+                1. Let |argumentsList| be a new [=/list=] with a single entry [=list/contain|containing=] |urlList|.
+                1. If |options| [=map/contains=] |data|, [=list/append=] |data| to |argumentsList|.
+                1. Let |operationResult| be the result of running [=Call=] on |operation| with |argumentsList|.
+                1. If |operationResult| has any error(s), then [=queue a global task=] on the [=DOM manipulation task source=], given |realm|'s [=global object=], to reject |promise| with a {{TypeError}}.
```
"has any error(s)" doesn't really mean anything unfortunately. I think you either need to [invoke this via Web IDL](https://webidl.spec.whatwg.org/#invoke-a-callback-function) and see if [an exception was thrown](https://webidl.spec.whatwg.org/#an-exception-was-thrown), or invoke this via TC39 primitives which you are doing (but I don't think you need to) and check for an abrupt completion record: https://wicg.github.io/turtledove/#:~:text=If%20result%20is%20an%20abrupt%20completion%2C%20return%20%C2%AB%20%22null%22%2C%20null%2C%20null%2C%20null%20%C2%BB

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+                1. Let |operationResult| be the result of running [=Call=] on |operation| with |argumentsList|.
+                1. If |operationResult| has any error(s), then [=queue a global task=] on the [=DOM manipulation task source=], given |realm|'s [=global object=], to reject |promise| with a {{TypeError}}.
+                1. Otherwise:
+                    1. Set |index| to the result of [=casting=] |operationResult| to an {{unsigned long}}.
```
Is there any reason you're using TC39 spec prose here instead of Web IDL conversions? https://webidl.spec.whatwg.org/#js-to-interface

xyaoinum replied:
Switched to Web IDL conversions. PTAL. (This is from the initial spec draft and I don't remember if there's a specific reason. cc @pythagoraskitty ).

domfarolino replied:
Superseded by https://github.com/WICG/shared-storage/pull/131#discussion_r1524964718.

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+                1. If |operationResult| has any error(s), then [=queue a global task=] on the [=DOM manipulation task source=], given |realm|'s [=global object=], to reject |promise| with a {{TypeError}}.
+                1. Otherwise:
+                    1. Set |index| to the result of [=casting=] |operationResult| to an {{unsigned long}}.
+                    1. If this throws an exception:
```
Please either use the Web IDL link I mentioned above to check for exceptions or explicitly check for an abrupt completion record.

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+                1. Otherwise:
+                    1. Set |index| to the result of [=casting=] |operationResult| to an {{unsigned long}}.
+                    1. If this throws an exception:
+                        1. Catch it and [=queue a global task=] on the [=DOM manipulation task source=], given |realm|'s [=global object=], to reject |promise| with a {{TypeError}}.
```
`[=reject=]`

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+                        1. Catch it and [=queue a global task=] on the [=DOM manipulation task source=], given |realm|'s [=global object=], to reject |promise| with a {{TypeError}}.
+                        1. Abort these steps.
+                    1. Otherwise, if |index| is greater than |urlList|'s [=list/size=]:
+                        1. [=Queue a global task=] on the [=DOM manipulation task source=], given |realm|'s [=global object=], to reject |promise| with a {{TypeError}}.
```
`[=reject=]`

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+                    1. Otherwise, if |index| is greater than |urlList|'s [=list/size=]:
+                        1. [=Queue a global task=] on the [=DOM manipulation task source=], given |realm|'s [=global object=], to reject |promise| with a {{TypeError}}.
+                        1. Abort these steps.
+                    1. [=Queue a global task=] on the [=DOM manipulation task source=], given |realm|'s [=global object=], to resolve |promise| with |index|.
```
`[=resolve=]`

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. Let |agent| be the result of [=obtaining a worklet agent=] given |outsideSettings|.
+    1. Run the following steps in |agent|:
+        1. Let |index| be [=default index=].
+        1. If |worklet|'s [=module map=] is not [=map/empty=]:
```
I'm a little confused about the workflow here. We just barely created the _brand new_ global at this point, so how could its operation map be non-empty at this point? Don't we have to let the script run and let it register its functions? At least I hope we're not trying to store any Web IDL callback function objects in the operation map and reuse them among globals... that seems bad. Real Worklets in HTML seem to store the module script response in the response map, but then only register the "name" of each operation. Then when an operation is run, we use the realm-agnostic name string to pluck the actual function off of the specific global we've selected to execute within, and then run that realm-bound function to execute the operation in the worklet.

xyaoinum replied:
[addModule](https://html.spec.whatwg.org/multipage/worklets.html#dom-worklet-addmodule)() (or createWorklet()) internally should invoke the script, which could invoke the [register()](file:///Users/yaoxia/Desktop/spec.html#dom-sharedstorageworkletglobalscope-register) method in the global scope.

This should be similar to [animation worklet’s registerAnimator](https://www.w3.org/TR/css-animation-worklet-1/#dom-animationworkletglobalscope-registeranimator).

domfarolino replied:
OK, and we know that this `addModule()` has always been called by this point because `selectURL()` throws a TypeError before this point if the list of global scopes is empty? That's the idea, right?

So in that case, I guess my question is now the opposite: how could the module map ever be _empty_ at this point? I'm assuming if we ever get this far, then we already have a worker module script that has loaded successfully, or is that wrong?

domfarolino replied:
Seems like this can be resolved given https://github.com/WICG/shared-storage/pull/131#issuecomment-1996109464.

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
-    1. If the result of running [=determine whether shared storage is allowed=] on the [=relevant settings object=] of [=this=] is false, return false.
-    1. If |worklet|'s [=addModule initiated=] is true, return false.
+    To <dfn>check if addModule is allowed and update status</dfn> given a {{SharedStorageWorklet}} |worklet| and an [=/URL=] |moduleURL|, run the following steps:
+    1. Let |workletOrigin| be |moduleURL|'s [=url/origin=].
```
This is interesting, but perhaps a little bit worrying. I want to check something, but let me know if I am missing something. At some point in the worklet creation process we need to call https://html.spec.whatwg.org/C#set-up-a-worklet-environment-settings-object, which actually sets up the global and its settings object appropriately. This is responsible for setting the global's origin, which is presumably what you _actually_ want to set `|workletOrigin|` to, right? Is this ever being done outside of this algorithm, or are we just missing it?

xyaoinum replied:
It seems the worklet will set it to the expected value during [addModule](https://html.spec.whatwg.org/multipage/worklets.html#dom-worklet-addmodule)() during its "Create a worklet global scope" step, so I assume there's no need to additionally mention it in this spec?

This `workletOrigin` member may seem redundant. But at this point, we do check for user preferences at the early stage of addModule(), so it also seems okay to have this member explicitly for that check (and for others checks as a convenience).

domfarolino replied:
OK so if I understand correctly, the intention is for https://pr-preview.s3.amazonaws.com/WICG/shared-storage/pull/131.html#sharedstorageworklet-worklet-origin to match, identically, the ESO's origin? If that's the case, can you add a note about this underneath the definition?

----

Also I want to point out one thing:

> 1. Let |workletOrigin| be |moduleURL|'s [=url/origin=].

Note that this is happening too early. The URL has been parsed yet, so `moduleURL` is just a string and doesn't have the concept of an "origin". This has to run _after_ this line: https://html.spec.whatwg.org/multipage/worklets.html#dom-worklet-addmodule:~:text=If%20moduleURLRecord%20is%20failure%2C%20then%20return%20a%20promise%20rejected%20with%20a%20%22SyntaxError%22%20DOMException

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. If [=this=]'s [=addModule initiated=] is false, then return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |window| be [=current global object=].
+    1. Assert that |window| is a {{Window}}.
+    1. Let |context| be |window|'s [=Window/browsing context=].
```
Similar comments from the above algorithm apply here...

domfarolino reviewed (changes requested):
Got through some more. I'll probably wait until the existing comments get resolved to continue.

xyaoinum commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
 Because adding multiple [=module scripts=] via {{Worklet/addModule()}} for the same {{SharedStorageWorklet}} would give the caller the ability to store data from [=Shared Storage=] in global variables defined in the [=module scripts=] and then exfiltrate the data through later call(s) to {{Worklet/addModule()}}, each {{SharedStorageWorklet}} can only call {{Worklet/addModule()}} once. The [=addMo ...
 
-When {{Worklet/addModule()}} is called for a worklet, it will run [=check if addModule is allowed and update status=], and if the result is false, abort the remaining steps in the {{Worklet/addModule()}} call, as detailed in the [[#worklet-monkey-patch]].
+When {{Worklet/addModule()}} is called for a worklet, it will run [=check if addModule is allowed and update status=], and if the result is "DisallowedDueToNonPreferenceError", or if the result is "DisallowedDueToPreferenceError" and the worklet's [=SharedStorageWorklet/is cross-origin worklet=] is false, abort the remaining steps in the {{Worklet/addModule()}} call, as detailed in the [[#worklet ...
```
I simplified it a bit. Not sure if it captures what you were asking. (e.g. did you intend to instead say _this will cause addModule to **~not~** fail_?)

domfarolino replied:
I think it's alright. I'm not sure I see the necessity in this paragraph in general, but I don't have strong opinions against it I guess.

xyaoinum commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+  <div class="note">
+    Here are the scenarios where the algorithms [=determine whether shared storage is allowed by context=] and [=determine whether shared storage is allowed by enrollment and user preference=] could be used:
+
+    - For <a href="#window-setter">Window setter/deleter methods</a>, |environment|'s [=environment settings object/origin=] should equal |origin|.
```
Reworded to "For [...], |environment| is [...], and |origin| is [...]". WDYT?

xyaoinum commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+    To <dfn>check if addModule is allowed and update status</dfn> given a {{SharedStorageWorklet}} |worklet| and an [=/URL=] |moduleURL|, run the following steps:
+    1. Let |workletOrigin| be |moduleURL|'s [=url/origin=].
+    1. If the result of running [=determine whether shared storage is allowed by context=] given [=this=]'s [=relevant settings object=] and |workletOrigin| is false, return "DisallowedDueToNonPreferenceError".
+    1. If |worklet|'s [=addModule initiated=] is true, return "DisallowedDueToNonPreferenceError".
```
It can be set if the algorithm (addModule()) was invoked on the same worklet object. Yes, the [=SharedStorageWorklet/worklet origin=] could also be set at this point.

domfarolino replied:
`|workletOrigin|` could be null here though, which breaks the call to "determine whether SS is allowed by context", which expects an origin (not null). So can you:

1. Only conditionally call this if `workletOrigin` is non-null
2. In the case where it is non-null, assert that it equals the [worklet origin member](https://pr-preview.s3.amazonaws.com/WICG/shared-storage/pull/131.html#sharedstorageworklet-worklet-origin)
    - Or would this assert be invalid? I guess maybe it is invalid because you can call `addModule()` several times with different (cross-origin) URLs?

?

domfarolino replied:
Ping on this.

domfarolino replied:
I just saw https://github.com/WICG/shared-storage/pull/131#issuecomment-1972259132. I'll respond there.

xyaoinum commented:
Thanks @domfarolino for the detailed review. I've addressed / responded to your comments. PTAL again. Thanks!

domfarolino commented on spec.bs:
```diff
@@ -201,68 +203,247 @@ This document introduces a new storage API that is intentionally not partitioned
-Each {{SharedStorageWorklet}} has an associated boolean <dfn>addModule initiated</dfn>, initialized to false.
+Each {{SharedStorageWorklet}} has an associated boolean <dfn for="SharedStorageWorklet">addModule initiated</dfn>, initialized to false.
+
+Each {{SharedStorageWorklet}} has an associated [=/origin=] <dfn for="SharedStorageWorklet">worklet origin</dfn>, initialized to null. It will be set to the |moduleURL|'s [=url/origin=] when {{Worklet/addModule()|addModule}}(|moduleURL|, <var ignore=''>options</var>) is called.
```
```suggestion
Each {{SharedStorageWorklet}} has an associated [=/origin=]-or-null <dfn for="SharedStorageWorklet">worklet origin</dfn>, initialized to null.

Note: It will be set to the worklet script's [=URL/origin=] during {{Worklet/addModule()}}.
```

domfarolino commented on spec.bs:
```diff
@@ -201,68 +203,247 @@ This document introduces a new storage API that is intentionally not partitioned
     2. If |pendingTasks| is 0, perform the following steps:
         1. If |workletGlobalScope| has an associated boolean [=addModule success=], set |workletGlobalScope|'s [=addModule success=] to true.
-        2. Resolve |promise|.
+        2. [=resolve=] |promise|.
```
```suggestion
        2. [=Resolve=] |promise|.
```

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
 
+  ## {{SharedStorageUrlWithMetadata}} and Reporting ## {#reporting}
+
+  A {{SharedStorageUrlWithMetadata}} {{/object}} is a [=dictionary=] containing a [=string=] representing a [=/URL=] and, optionally, a {{reportingMetadata}} {{/object}}.
```
```suggestion
  {{SharedStorageUrlWithMetadata}} [=dictionaries=] contain a [=string=] representing a [=/URL=] and, optionally, a {{reportingMetadata}} {{/object}}.
```

SSUWM is not an object, it is _just_ a dictionary. There are a few more references below that reference SSUWM as an `{{/object}}`, can you get rid of those?

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  If a {{SharedStorageUrlWithMetadata}} {{/object}} contains a non-[=map/empty=] {{SharedStorageUrlWithMetadata/reportingMetadata}} {{/object}} in the form of a [=dictionary=] whose [=map/keys=] are [=FenceEvent/eventTypes=] and whose [=map/values=] are [=strings=] that parse to valid [=/URLs=], then these [=FenceEvent/eventType=]-[=/URL=] pairs will be [=register reporting metadata|registered=]  ...
+
+  Inside a [=fenced frame=] with [=FenceEvent/eventType=]-[=/URL=] pairs that have been [=register reporting metadata|registered=] through {{SharedStorageWorklet/selectURL()}} with {{SharedStorageUrlWithMetadata/reportingMetadata}} {{/object}}s, if [=fence.reportEvent()=] is called on a [=FenceEvent=] with a [=FenceEvent/destination=] [=list/containing=] "`shared-storage-select-url`" and that [=F ...
```
All of the `[=FenceEvent=]` stuff should be `{{FenceEvent}}`.

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+    To <dfn>validate reporting metadata</dfn>, given an {{/object}} |reportingMetadata|, run the following steps:
+
+    1. If |reportingMetadata| is [=map/empty=], return true.
+    1. If |reportingMetadata| is not a [=dictionary=], return false.
```
These first two steps seem like they should be switched around?

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+    1. If |reportingMetadata| is [=map/empty=], return true.
+    1. If |reportingMetadata| is not a [=dictionary=], return false.
+    1. [=map/iterate|For each=] <var ignore="">eventType</var> -> |urlString| of |reportingMetadata|, if the result of running [=get the canonical URL string if valid=] with |urlString| is undefined, return false.
```
```suggestion
    1. [=map/iterate|For each=] <var ignore="">eventType</var> → |urlString| of |reportingMetadata|, if the result of running [=get the canonical URL string if valid=] with |urlString| is undefined, return false.
```

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+  </div>
+
+  <div algorithm>
+    To <dfn>register reporting metadata</dfn>, given an {{/object}} |reportingMetadata| and a "fenced frame config struct" |fencedFrameConfigStruct|, run the following steps:
```
> "fenced frame config struct"

Link to the actual thing?

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+    1. If |reportingMetadata| is [=map/empty=], return.
+    1. [=Assert=]: |reportingMetadata| is a [=dictionary=].
+    1. Let |reportingUrlMap| be an [=map/empty=] [=map=].
+    1. [=map/iterate|For each=] |eventType| -> |urlString| of |reportingMetadata|:
```
```suggestion
    1. [=map/iterate|For each=] |eventType| → |urlString| of |reportingMetadata|:
```

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+        1. Let |url| be the result of running a [=URL parser=] on |urlString|.
+        1. [=Assert=]: |url| is a valid [=/URL=].
+        1. [=map/Set=] |reportingUrlMap|[|eventType|] to |url|.
+    1. <span class=todo>Store |reportingUrlMap| inside a "fenced frame reporter" class associated with |fencedFrameConfigStruct|. Both of these still need to be added to the draft [=fenced frame|Fenced Frame specification=].</span>
```
Link to the right term reporter dfn?

domfarolino replied:
```suggestion
    1. <span class=todo>Store |reportingUrlMap| inside a "fenced frame reporter" class associated with |fencedFrameConfigStruct|. Both of these still need to be added to the draft [[Fenced-Frame]].</span>
```

You want a biblio link like that. See https://github.com/WICG/fenced-frame/blob/master/spec.bs#L361 for example. You might have to mess around with the custom biblio, I'm really not sure: https://github.com/WICG/fenced-frame/blob/master/spec.bs#L29-L39

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  ### Navigation Entropy Budget ### {#nav-budget}
+
+  If a user activates a [=fenced frame=] whose {{FencedFrameConfig}} was generated by {{SharedStorageWorklet/selectURL()}} and thereby initiates a [=top-frame=] [=navigate|navigation=], this will reveal to the landing page that its [=/URL=] was selected, which is a leak in [=entropy bits=] of up to logarithm base 2 of the number of input [=/URLs=] for the call to {{SharedStorageWorklet/selectURL( ...
```
It's unfortunately hard to suggest specific changes to massive massive blocks like this since they are not wrapped at all :(

But you should link "activates" to HTML, like we do in fenced frame: https://github.com/WICG/fenced-frame/blob/master/spec.bs#L167

domfarolino replied:
```suggestion
  If a user activates a [=fenced frame=] whose {{FencedFrameConfig}} was generated by {{SharedStorageWorklet/selectURL()}} and thereby initiates a [=top-level traversable=] [=navigate|navigation=], this will reveal to the landing page that its [=/URL=] was selected, which is a leak in [=entropy bits=] of up to logarithm base 2 of the number of input [=/URLs=] for the call to {{SharedStorageWorklet/selectURL()}}. To mitigate this, a [=user agent=] will set a per-[=calling site=] [=navigation entropy allowance=].
```

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  If a user activates a [=fenced frame=] whose {{FencedFrameConfig}} was generated by {{SharedStorageWorklet/selectURL()}} and thereby initiates a [=top-frame=] [=navigate|navigation=], this will reveal to the landing page that its [=/URL=] was selected, which is a leak in [=entropy bits=] of up to logarithm base 2 of the number of input [=/URLs=] for the call to {{SharedStorageWorklet/selectURL( ...
+
+  A <dfn>calling site</dfn> for {{SharedStorageWorklet/selectURL()}} is the [=site=] resulting from running [=obtain a site=] with the [=url/origin=] of an [=environment=] that makes a {{SharedStorageWorklet/selectURL()}} call.
```
```suggestion
  A <dfn>calling site</dfn> for {{SharedStorageWorklet/selectURL()}} is the [=site=] resulting from running [=obtain a site=] with the [=url/origin=] of an [=environment=] that makes a {{SharedStorageWorklet/selectURL()}} call.
```
```suggestion
  A <dfn>calling site</dfn> for {{SharedStorageWorklet/selectURL()}} is a [=site=].
  
  Note: It is the result of running [=obtain a site=] with the [=url/origin=] of an [=environment=] that makes a {{SharedStorageWorklet/selectURL()}} call.
```

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  An <dfn>navigation entropy ledger</dfn> is a [=/list=] of [=bit debits=].
+
+  A <dfn>bit debit</dfn> is a [=struct=] containing a {{double}} <dfn for="bit debit">bits</dfn>, indicating a value in [=entropy bits=], along with a {{DOMHighResTimeStamp}} <dfn for="bit debit">timestamp</dfn> (from the [=Unix Epoch=]).
```
I think laying this out to be more readable would be best. See https://github.com/WICG/fenced-frame/blob/master/spec.bs#L657-L665 for example.

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  {{SharedStorageWorklet/selectURL()}}'s argument "`urls`" is its <dfn for=selectURL>input URL list</dfn>.
+
+  When a [=calling site=] has insufficient [=calling site/remaining navigation budget=], {{SharedStorageWorklet/selectURL()}} will return a {{SharedStorageResponse}} (i.e. either a {{FencedFrameConfig}} or an opaque [=/URL=]) for the {{SharedStorageUrlWithMetadata/url}} in the {{SharedStorageUrlWithMetadata}} at the [=default index=] in its [=selectURL/input URL list=].
```
What's an "opaque URL"? Do you mean [a URN](https://wicg.github.io/fenced-frame/#ref-for-urn-uuid%E2%91%A8)?

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  When a [=calling site=] has insufficient [=calling site/remaining navigation budget=], {{SharedStorageWorklet/selectURL()}} will return a {{SharedStorageResponse}} (i.e. either a {{FencedFrameConfig}} or an opaque [=/URL=]) for the {{SharedStorageUrlWithMetadata/url}} in the {{SharedStorageUrlWithMetadata}} at the [=default index=] in its [=selectURL/input URL list=].
+
+  The <dfn>default index</dfn> for a call to {{SharedStorageWorklet/selectURL()}} is implementation-defined in such a way that it is independent from the result of the associated {{SharedStorageSelectURLOperation}}'s "`run`" method.
```
Link https://infra.spec.whatwg.org/#implementation-defined

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+    1. Otherwise, return false.
+  </div>
+
+  A [=bit debit=] will need to be [=charge shared storage navigation budget|charged=] to the [=shared storage navigation budget table=] for each [=top-frame=] [=navigate|navigation=] initiated by a [=fenced frame=] whose {{FencedFrameConfig}} was generated via {{SharedStorageWorklet/selectURL()}}, as this can leak cross-site data. Since the [=bit debit/bits=] to charge is calculated during the ca ...
```
```suggestion
  A [=bit debit=] will need to be [=charge shared storage navigation budget|charged=] to the [=shared storage navigation budget table=] for each [=top-frame=] [=navigate|navigation=] initiated by a [=fenced frame=] whose {{FencedFrameConfig}} was generated via {{SharedStorageWorklet/selectURL()}}, as this can leak cross-site data. Since the [=bit debit/bits=] to charge is calculated during the call to {{SharedStorageWorklet/selectURL()}} but only actually recorded in the [=shared storage navigation budget table=] if and when the resulting fenced frame initiates a [=top-level traversable=] [=beginning navigation|navigation=], the [=bit debit/bits=] must be stored as a <dfn>pending shared storage budget debit</dfn> in the corresponding {{FencedFrameConfig}} until this time.
```

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  Between [=beginning navigation=] and [=ending navigation=], a [=user agent=] will perform the [=charge shared storage navigation budget=] algorithm.
+
+  Issue: The "fenced frame config struct" and its boolean <dfn>has navigated</dfn> have not yet been added to the draft [=Fenced Frame=] specification. Some form of them will be added, although their names are subject to bikeshedding. <span class=todo>Fix the names when they are added.</span>
```
Link: "fenced frame config struct"

Link: [=Fenced Frame=]

xyaoinum replied:
I linked the "fenced frame config struct". But WDYM by linking [=Fenced Frame=]? It was already there.

domfarolino replied:
I think I meant `<{fencedframe}>`

xyaoinum replied:
On second look, I seems I should probably link to `[[Fenced-Frame]]` rather than the `<fencedframe>` element?

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  A [=bit debit=] will need to be [=charge shared storage navigation budget|charged=] to the [=shared storage navigation budget table=] for each [=top-frame=] [=navigate|navigation=] initiated by a [=fenced frame=] whose {{FencedFrameConfig}} was generated via {{SharedStorageWorklet/selectURL()}}, as this can leak cross-site data. Since the [=bit debit/bits=] to charge is calculated during the ca ...
+
+  Between [=beginning navigation=] and [=ending navigation=], a [=user agent=] will perform the [=charge shared storage navigation budget=] algorithm.
```
When? Where do we actually call this? Unless I'm missing it, you'll need to actually call this at the right time during the navigation flow. We can possibly follow-up in a separate PR to figure out the right timing for this, but we should at the very least think this through and track it with a bug if it's not already done.

xyaoinum replied:
Ack. Given that this isn't relevant to this patch (this just move sections around, due to the fact that we moved the central algorithm of selectURL()/run() to under the `SharedStorageWorklet` interface, from the `SharedStorage` interface), created https://github.com/WICG/shared-storage/issues/138 to track.

xyaoinum replied:
Done. Linked Issue(138).

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+  Issue: The "fenced frame config struct" and its boolean <dfn>has navigated</dfn> have not yet been added to the draft [=Fenced Frame=] specification. Some form of them will be added, although their names are subject to bikeshedding. <span class=todo>Fix the names when they are added.</span>
+
+  <div algorithm>
+    To <dfn>charge shared storage navigation budget</dfn> during a [=beginning navigation|navigation=] with [=/navigable=] |navigable| and {{Document}} |sourceDocument|, run the following steps:
```
I don't think we should specify "during a navigation"... we should just monkeypatch HTML to actually call this at the right time.

xyaoinum replied:
Ack. Created https://github.com/WICG/shared-storage/issues/138 to track.

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+  <div algorithm>
+    To <dfn>charge shared storage navigation budget</dfn> during a [=beginning navigation|navigation=] with [=/navigable=] |navigable| and {{Document}} |sourceDocument|, run the following steps:
+
+    1. If |navigable| is not a [=navigable/traversable navigable=], return.
```
traversable navigable might be fine, but do we need to specify top-level traversable instead?

domfarolino commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+    To <dfn>charge shared storage navigation budget</dfn> during a [=beginning navigation|navigation=] with [=/navigable=] |navigable| and {{Document}} |sourceDocument|, run the following steps:
+
+    1. If |navigable| is not a [=navigable/traversable navigable=], return.
+    1. Let |node| be |sourceDocument|'s [=node navigable=].
```
`|node|` is already a common word for most DOM things that aren't navigables. Can you rename this variable to something like `|current navigable|` or something?

xyaoinum commented on spec.bs:
```diff
@@ -201,68 +203,247 @@ This document introduces a new storage API that is intentionally not partitioned
+  <div class="note">
+    Here are the scenarios where the algorithms [=determine whether shared storage is allowed by context=] and [=determine whether shared storage is allowed by enrollment and user preference=] are used:
+
+    - For [[#window-setter]], |environment| is the current context, and |origin| is |environment|'s [=environment settings object/origin=].
```
Updated to `For each method under [[#window-setter]], ...`

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +530,196 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+  <xmp class='idl'>
+    dictionary SharedStorageUrlWithMetadata {
+      required USVString url;
+      object reportingMetadata;
```
Not that I'm aware of. cc-ing @pythagoraskitty  who authored the change.

xyaoinum replied:
@pythagoraskitty could you take a look? i.e. Is there a reason `reportingMetadata` is an `object` instead of [any](https://webidl.spec.whatwg.org/#idl-any)?

pythagoraskitty replied:
Well, `reportingMetadata` needs to be a dictionary, so the Chromium implementation currently assumes it's an `object`: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/shared_storage/shared_storage_worklet.cc;drc=4ec42762cc8329e274218f13828754b17941c2ff;l=3

domfarolino replied:
If reporting metadata is supposed to be a dictionary, then it should have its own dictionary type (just like `SharedStorageUrlWithMetadata`) and just be embedded inside `SharedStorageUrlWithMetadata` as a member.

xyaoinum replied:
I think it can be a dictionary / don't see any reason why it cannot be. Created issue https://github.com/WICG/shared-storage/issues/141 to track, so that @pythagoraskitty can respond and/or address this separately.

xyaoinum commented:
@domfarolino For comment https://github.com/WICG/shared-storage/pull/131#discussion_r1508141219, strangely, there's no way to respond, so I'm writing my response here:

`|workletOrigin|` is a valid URL's (i.e. `|moduleURLRecord|`) [origin](https://url.spec.whatwg.org/#concept-url-origin). How can it be null? Do you mean opaque? In that case, opaque origin *can* be handled by "determine whether shared storage is allowed by context".

Also, I think it might make it more readable to put the [=addModule initiated=] check to be front of the algorithm, and assert that `|worklet|'s [=worklet origin=]` is null afterwards. WDYT?

xyaoinum commented:
@domfarolino I addressed you last comments. PTAL again. Thanks!

domfarolino commented on spec.bs:
```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. If |environment|'s [=associated document=] is not [=fully active=], return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |realm| be the [=current realm=].
+    1. Let |outsideSettings| be |worklet|'s [=relevant settings object=].
```
I'm not sure if you agree or disagree since this was marked as resolved without comment. WDYT?

xyaoinum replied:
I addressed it directly in the spec:
I changed `|environment|'s [=associated document=]` to `|window|'s [=associated document=]` per your other comment, so the `|environment|` variable is gone now. Thus, here, I'll need to keep `|worklet|'s [=relevant settings object=]`.

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. If [=this=]'s [=global scopes|list of global scopes=] is [=list/empty=], then return a [=promise rejected=] with a {{TypeError}}.
+    1. [=Assert=]: [=this=]'s [=global scopes|list of global scopes=] [=list/contains=] a single {{SharedStorageWorkletGlobalScope}}.
+    1. If the result of running [=SharedStorageWorkletGlobalScope/check whether addModule is finished=] for [=this=]'s {{SharedStorageWorkletGlobalScope}} is false, return a [=promise rejected=]  with a {{TypeError}}.
+    1. If |urls| is empty or exceeds the maximum allowed length, return a [=promise rejected=] with a {{TypeError}}.
```
You have to define the length somewhere.

xyaoinum replied:
I wrote down the number (8) inline.

domfarolino replied:
OK. It might be nice to provide some documentation, just as you would in code maybe, as to why that number was chosen etc. Up to you though I guess.

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+    4. If |this| is of type {{SharedStorageWorklet}}:
+        1. Let |addModuleAllowedResult| be the result of running [=check if addModule is allowed and update state=] given |this| and <var ignore=''>moduleURLRecord</var>.
+        1. If |addModuleAllowedResult| is "DisallowedDueToNonPreferenceError":
+            1. Reject <var ignore="">promise</var> with a {{TypeError}}.
```
Link to reject and resolve throughout this whole algorithm. It doesn't look like there are any references to those terms.

domfarolino replied:
Since you use promise multiple times, you don't have to ignore it as a variable in any instance.

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+        1. Else:
+            1. [=Assert=]: |addModuleAllowedResult| is "Allowed".
+            1. Resolve <var ignore="">promise</var>.
+        1. Return <var ignore="">promise</var>.
```
If I have asked this before, them I'm sorry. Do we really always unconditionally return a promise right away in the `SharedStorageWorklet` case? It looks like we'd never run the rest of the steps in https://html.spec.whatwg.org/#dom-worklet-addmodule, and therefore we'd never fetch the module script or create the global (and store it to the list of global scopes) or anything. Am I missing something?

This is kind of related to a concern I have about calling the "obtain a worklet agent" algorithm manually in both of the select and run helper algorithms. Calling that algorithm, I think, will create a fresh worklet agent from scratch each time. This agent/realm/execution context trio will be completely different from the agent associated with the SINGLE global scope that exists in the "list of global scopes" (for SharedStorageWorkletGlobalScope). Furthermore, we're storing functions from some(??) agent in the operations map, and then calling them in this new freshly-created agent, which is bad. Let me know if that doesn't make sense, or if I'm off-base somewhere.

domfarolino replied:
> and the global scope should be associated with a single [operation map](https://wicg.github.io/shared-storage/#sharedstorageworkletglobalscope-operation-map). And thus, the references in the spec seem legitimate in general.

This is very useful to know (and I probably knew it but forgot, sadly). Can you add an assertion to every place where we pull an IDL callback function out of the operations map and invoke it, to assert that the function object's [associated realm](https://dontcallmedom.github.io/webdex/a.html#associated%20realm%40%40webidl%25%25dfn) (see https://webidl.spec.whatwg.org/#dfn-associated-realm) matches the [relevant realm](https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-realm) of the global object?

> can different agents access the same global scope (from the spec's perspective)?

Hmm, I'm curious: why is this question relevant? It sounds like there is only ever a single shared storage worklet agent created for each hosting document, right?

xyaoinum replied:
Added the assert.

> Hmm, I'm curious: why is this question relevant?

The [obtain a worker/worklet agent](https://html.spec.whatwg.org/multipage/webappapis.html#obtaining-a-worker/worklet-agent) algorithm creates an agent each time. So it seems it each worklet corresponds to a single "agent cluster", but there could be multiple "agents".

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. If the result of running [=Is feature enabled in document for origin?=] on "[=PermissionsPolicy/shared-storage-select-url=]", |document|, and [=this=]'s [=SharedStorageWorklet/worklet origin=] returns false, return a [=promise rejected=] with a {{TypeError}}.
+    1. If [=this=]'s [=global scopes|list of global scopes=] is [=list/empty=], then return a [=promise rejected=] with a {{TypeError}}.
+    1. [=Assert=]: [=this=]'s [=global scopes|list of global scopes=] [=list/contains=] a single {{SharedStorageWorkletGlobalScope}}.
+    1. If the result of running [=SharedStorageWorkletGlobalScope/check whether addModule is finished=] for [=this=]'s {{SharedStorageWorkletGlobalScope}} is false, return a [=promise rejected=]  with a {{TypeError}}.
```
```suggestion
    1. If the result of running [=SharedStorageWorkletGlobalScope/check whether addModule is finished=] for [=this=]'s {{SharedStorageWorkletGlobalScope}} is false, return a [=promise rejected=] with a {{TypeError}}.
```

domfarolino replied:
"this's SharedStorageWorkerGlobalScope" is not a real member. Do you mean to grab the global scope from the list of global scopes, associated with the outer worklet class (that hangs off of `Window`)?

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+    To <dfn>get the select-url result index</dfn>, given {{SharedStorageWorklet}} |worklet|, {{DOMString}} |operationName|, [=/list=] of {{SharedStorageUrlWithMetadata}}s |urlList|, and {{SharedStorageRunOperationMethodOptions}} |options|:
+
+    1. Let |promise| be a new [=promise=].
+    1. Let |window| be the [=current global object=].
```
The "current" dfns are kinda tricky. It should be the same global as `|worklet|'s relevant settings object` I think, so I'd probably use that just to avoid confusion between what global we're trying to work with, and what global `|worklet|` comes from. This makes it clearer that we're operating with a single global, and makes it less likely that somebody thinks we have to use `[=current global object=]` intentionally, because the globals could be different.

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. Return |promise|, and immediately [=obtain a worklet agent=] given |outsideSettings| and run the rest of these steps in that agent:
+        1. Let |index| be [=default index=].
+        1. If |worklet|'s [=module map=] is not [=map/empty=]:
+            1. Let |operationMap| be the associated {{SharedStorageWorkletGlobalScope}}'s [=SharedStorageWorkletGlobalScope/operation map=].
```
It's not clear what "the associated" means. I think you are referring to the global scope in the "list of global scopes", of which there can only be one. If my understanding is right, can you pull it out in a separate variable?

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+        1. Let |index| be [=default index=].
+        1. If |worklet|'s [=module map=] is not [=map/empty=]:
+            1. Let |operationMap| be the associated {{SharedStorageWorkletGlobalScope}}'s [=SharedStorageWorkletGlobalScope/operation map=].
+            1. If |operationMap| [=map/contains=] |operationName|:
```
When would `|operationMap|` NOT contain `|operationName|`? Would that be if `register()` was never called, or was called with a different name? If so, can you just maybe document that here with a `Note:` below or something?

There are a lot of conditions in this function, so I think it would be good to document the many edge cases to illustrate how they can be hit and why / when they are valid / invalid.

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+        1. Else, [=resolve=] |promise| with undefined.
+        1. Return |promise|.
+    1. Return |promise|, and immediately [=obtaining a worklet agent=] given |outsideSettings| and run the rest of these steps in that agent:
+        1. [=Queue a global task=] on the [=DOM manipulation task source=], given |window|, to [=resolve=] |promise| with undefined.
```
Why do we immediately queue a task to resolve the returned promise? That means the promise resolution has nothing to do with the output or execution of the worklet function, right? This promise always just resolves as fast as it can (gated by a single task) and that's it? Maybe documenting that would be good with a `Note:`.

domfarolino replied:
Has this been resolved? Please leave comments on things that you resolved. Just reading the code again and trying to get back some context, I don't see this as resolved.

xyaoinum replied:
I did left a `Note:`: _"Note: The |promise|'s resolution should be before and not depend on the execution inside {{SharedStorageWorkletGlobalScope}}. This is because shared storage is a type of unpartitioned storage, and a {{SharedStorageWorkletGlobalScope}} can have access to cross-site data, which shouldn't be leaked via {{SharedStorageWorklet/run()}} (via its success/error result)."_

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+                1. Otherwise, [=invoke=] |operation| without any arguments list.
+        1. If |options|["`keepAlive`"] is false:
+            1. Wait for |operation| to finish running, if applicable.
+            1. Run [=terminate a worklet global scope=] with [=this=]'s {{SharedStorageWorkletGlobalScope}}.
```
`[=this=]` is weird here. We're running inside of a worklet global scope, so I think `this` refers to the global scope actually. Same with the above instance on 404. Does that make sense?

xyaoinum replied:
Yes. Changed to just [=this=].

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+
+        Issue: The [=fenced frame config=] and the following "obtain a {{FencedFrameConfig}} from a fenced frame config struct" algorithm have not yet been added to the draft [=Fenced Frame=] specification. Some form of them will be added, although their names are subject to bikeshedding.
+
+    1. If |options|["`resolveToConfig`"] is true, resolve |resultPromise| with the result of running "obtain a {{FencedFrameConfig}} from a fenced frame config struct" with |fencedFrameConfigStruct|. <span class=todo>Add correct struct and algorithms names as well as linking when Fenced Frame API updates their draft spec to include it.</span>
```
link to resolve and reject throughout this entire algorithm please.

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+        Issue: The [=fenced frame config=] and the following "obtain a {{FencedFrameConfig}} from a fenced frame config struct" algorithm have not yet been added to the draft [=Fenced Frame=] specification. Some form of them will be added, although their names are subject to bikeshedding.
+
+    1. If |options|["`resolveToConfig`"] is true, resolve |resultPromise| with the result of running "obtain a {{FencedFrameConfig}} from a fenced frame config struct" with |fencedFrameConfigStruct|. <span class=todo>Add correct struct and algorithms names as well as linking when Fenced Frame API updates their draft spec to include it.</span>
+    1. Othewise, resolve |resultPromise| to |fencedFrameConfigStruct|'s "urn uuid". <span class=todo>Add correct struct name and urn:uuid name as well as linking when Fenced Frame API updates their draft spec to include it.</span>
```
Should exist by now.

xyaoinum replied:
Hmm.. the internal struct type ([FencedFrameConfig](https://wicg.github.io/fenced-frame/#fenced-frame-config); the one we are using here) still _does not_ contain any urn uuid type. Only the JavaScript type ([FencedFrameConfig](https://wicg.github.io/fenced-frame/#fencedframeconfig)) has it. Anyway, I feel it could be done separately as there are a lot of other places referring to FF/TODO, and the change seems orthogonal to this patch.

domfarolino replied:
Well we should at least check with people active over on that specification to see if there are plans to address that TODO. This TODO (and others?) make sound like the FF spec is planning on adding a URN UUID member to the struct, but to my knowledge they are not. We should make the need for this more vocally. @gtanzer, what do you think? Can we help resolve this?

gtanzer replied:
I've never really understood in this:

"One major input to the fencedframe element is the FencedFrameConfig interface, which maps to an internal fenced frame config struct."

what exactly "maps to an internal fenced frame config struct" means for an interface in spec. It really maps to a pair of a urn uuid and a fenced frame config struct anyway.

Regardless, there are way more TODOs than are necessary. All of the stuff about pending configs / generating urns uses the interface in fenced rame config mapping.
https://wicg.github.io/fenced-frame/#fenced-frame-config-mapping

e.g. here in the PA spec:
https://wicg.github.io/turtledove/#:~:text=Let%20fencedFrameConfig%20be%20the%20result%20of%20filling%20in%20a%20pending%20fenced%20frame%20config%20with%20pendingConfig%2C%20auctionConfig%2C%20winnerInfo%2C%20and%20auctionReportInfo.

which is a bit sloppy with the urn part.

I think the issue is that originally we intended not to use urn:uuids in spec, since they are only necessary in implementation as a browser/renderer thing. In spec you can just use references to the same underlying config object.

But urn iframes complicate everything, and then I think a bunch of stuff is spec'd in terms of lookups into the config mapping using the urn...

domfarolino commented on spec.bs:
```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. If |fencedFrameConfigStruct|'s "urn uuid" is invalid (i.e. instead of pending mapped), then return. <span class=todo>Add correct struct name and urn:uuid name as well as linking when Fenced Frame API updates their draft spec to include it.</span>
+    1. Let |indexPromise| be the result of running [=get the select-url result index=], given [=this=], |name|, |urlList|, and |options|.
+    1. [=Upon fulfillment=] of |indexPromise|, perform the following steps:
+        1. Let |resultIndex| be the numerical value of |indexPromise|.
```
I'm not sure what this means. Promises don't have a numerical value. Do you mean to get the resolved value of the promise? In that case, follow an example like this (https://w3c.github.io/ServiceWorker/#:~:text=Upon%20fulfillment%20of%20r%20with%20response) where you define the value that the promise gets fultilled with, in the definition of the steps that act on it.

domfarolino commented:
Responding to https://github.com/WICG/shared-storage/pull/131#issuecomment-1972259132.

> |workletOrigin| is a valid URL's (i.e. |moduleURLRecord|) [origin](https://url.spec.whatwg.org/#concept-url-origin). How can it be null? Do you mean opaque? In that case, opaque origin can be handled by "determine whether shared storage is allowed by context".

Hmm, I can't remember exactly what I was thinking, but I probably was responding to the PR when we handled the URL manually before steps 1-3 in https://html.spec.whatwg.org/C#dom-worklet-addmodule did all of the URL parsing work.

https://github.com/WICG/shared-storage/blob/6606cc0e380b8ff1c2aa2d927c3dd3b03febadbb/spec.bs#L416-L419

In that case, the URL could be invalid and I guess a failure (but not null, you're right)? Either way, it looks OK now, since we're doing this _after_ HTML's `addModule()` gets a first crack at parsing the URL.

> Also, I think it might make it more readable to put the [=addModule initiated=] check to be front of the algorithm, and assert that |worklet|'s [=worklet origin=] is null afterwards. WDYT?

I think that sounds fine.

xyaoinum commented:
@domfarolino PTAL again. Thanks!

Also, responding to https://github.com/WICG/shared-storage/pull/131#discussion_r1520124784 here, as again, I see no way to respond to that comment inline: "re: I think you are right. The module map check is redundant. Switched to an [=Assert=]."

xyaoinum commented on spec.bs:
```diff
@@ -201,68 +210,264 @@ This document introduces a new storage API that is intentionally not partitioned
+            Note: This indicates that |operationCtor|'s run() method encounters an error, where |operationCtor| is the parameter in {{SharedStorageWorkletGlobalScope/register()}}.
+
+        1. Otherwise:
+            1. Set |index| to |operationResult|, [=converted to a JavaScript value=] of type {{unsigned long}}.
```
It is supposed to be a `run()` function inside the class (corresponding to the constructor). For example, inside the module script, it'd do
```
class SelectURLOperation {
  async run(data, urls) {
    return 3;
  }
}

register('select-url-for-experiment', SelectURLOperation);
```

In the [Chrome code](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/shared_storage/shared_storage_worklet_global_scope.cc?q=%22retriever.GetMethodOrThrow%22), we don't define any IDL, but rather we "manually" retrieve the the run() method. In the [spec](https://wicg.github.io/shared-storage/#dom-sharedstorageworkletglobalscope-register ), we have _"Let run be the result of running [GetMethod()](https://tc39.es/ecma262/#sec-getmethod) with prototype and "run"._

I believe something is still missing, but I don't know how to specify this properly. Should we define the type somewhere (via IDL)? Should we not do [=invoke=] here, but do something else in order to "call this function"?

xyaoinum replied:
Updated following the worklet example. Also removed the definition and references for those "Operation" types.

But the original concern remains: At the [=invoke=] time, as you said, it already returns an IDL type. But how it's specified to be {{unsigned long}} / how do I convert it to that type?

domfarolino replied:
> But the original concern remains: At the [=invoke=] time, as you said, it already returns an IDL type. But how it's specified to be {{unsigned long}} / how do I convert it to that type?

You'll make an IDL callback type like https://wicg.github.io/observable/#callbackdef-predicate, but instead of `bool` as the return type, you want `unsigned long` I think. You can choose whatever arguments as well. Then instead of:

> Let callback be the result of [converting](https://webidl.spec.whatwg.org/#es-type-mapping) function to a Web IDL [Function](https://webidl.spec.whatwg.org/#common-Function) instance.

You'll convert to a Web IDL "your callback type here", type.

xyaoinum replied:
Got it. Done. (Introduced callback `RunFunctionForSharedStorageSelectURLOperation` type, and updated the usage site.)

xyaoinum commented:
@domfarolino: I addressed / responded to you comments (other than the one waiting on @gtanzer). PTAL again. Thanks!

xyaoinum commented on spec.bs:
```diff
@@ -201,68 +210,266 @@ This document introduces a new storage API that is intentionally not partitioned
 
   <div algorithm>
-    To <dfn>determine whether shared storage is allowed</dfn>, given an [=environment settings object=] |environment|, run these steps:
+    To <dfn>determine whether shared storage is allowed by context</dfn>, given an [=environment settings object=] |environment| and an [=url/origin=] |origin|, run these steps:
```
Fixed.

xyaoinum commented on spec.bs:
```diff
@@ -201,68 +210,266 @@ This document introduces a new storage API that is intentionally not partitioned
 
 Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=], initially empty, can contain at most one instance of its [=worklet global scope type=], the {{SharedStorageWorkletGlobalScope}}.
 
+  Note: The website that serves the module script should be aware of the implication of CORS: when the the module script's [=/URL=]'s [=url/origin=] and the worklet's creator window's origin is not [=same origin=], and by granting the module script resource via CORS, it will also grant the worklet's creation and subsequent operations on the worklet, under module script's [=/URL=]'s [=url/origin=] ...
```
Done.

xyaoinum commented on spec.bs:
```diff
@@ -201,68 +210,266 @@ This document introduces a new storage API that is intentionally not partitioned
   ## The {{SharedStorageWorkletGlobalScope}} ## {#global-scope}
 
-  The {{SharedStorageWorklet}}'s [=worklet global scope type=] is {{SharedStorageWorkletGlobalScope}}. 
+  The {{SharedStorageWorklet}}'s [=worklet global scope type=] is {{SharedStorageWorkletGlobalScope}}.
```
Done.

xyaoinum commented on spec.bs:
```diff
@@ -289,6 +496,8 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
 
   Each {{SharedStorageWorkletGlobalScope}} has an associated [=environment settings object=] <dfn for=SharedStorageWorkletGlobalScope>outside settings</dfn>, which is the associated {{SharedStorageWorklet}}'s [=relevant settings object=].
 
+  Each {{SharedStorageWorkletGlobalScope}} has an associated [=url/origin=] <dfn for=SharedStorageWorkletGlobalScope>worklet origin</dfn>, which is initialized to the associated {{SharedStorageWorklet}}'s [=SharedStorageWorklet/worklet origin=].
```
Looking at this now, I don't think the worklet _owns_ that `insideSettings` object. IIUC, `insideSettings` is a temporary variable in the [create-a-worklet-global-scope](https://html.spec.whatwg.org/multipage/worklets.html#create-a-worklet-global-scope) algorithm, whereas for shared storage, we want to pin it to the worklet for future retrieval (e.g. access the data storage for that origin). So, I actually do not see a duplication here.

domfarolino replied:
The worklet global scope has a realm, and that realm owns the settings object, and that settings object has an origin: https://html.spec.whatwg.org/multipage/worklets.html#set-up-a-worklet-environment-settings-object:~:text=Set%20realm%27s%20%5B%5BHostDefined%5D%5D%20field%20to%20settingsObject. And that origin is opaque.

xyaoinum replied:
Got it. I removed all the new "worklet origin" fields and added a few more monkey patches. PTAL. (I did encounter an error when I attempted to refer to the realm's `[[HostDefined]]` field. I temporarily wrapped them with TODOs. Let me know if you know how to specify this properly.)

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +556,204 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+    1. Let |currentNavigable| be |sourceDocument|'s [=node navigable=].
+    1. While |currentNavigable| is not null:
+        1. Let |site| be the result of running [=obtain a site=] with |currentNavigable|'s [=active document=]'s [=document/origin=].
+        1. If |currentNavigable| has a [=fenced frame config=] and |site| is not an [=opaque origin=], perform the following steps:
```
I think the intent is to check "whether the navigation is initiated from a fenced frame". It also doesn't seem right to me but I don't know how to fix it right away (e.g. is "fenced frame config" a member of "navigable"?). I added this as a sub-task of the issue https://github.com/WICG/shared-storage/issues/138 I filed previously. Or, I can fix it now if @pythagoraskitty (the original author of this part) or @gtanzer knows the answer.

xyaoinum replied:
re the 2nd question: There's a last step in the end `Set |currentNavigable| to |currentNavigable|'s [=navigable/parent=]`, so we cannot simply early return.

gtanzer replied:
This is an example of how to check whether the current document is a fenced frame: 

https://wicg.github.io/fenced-frame/#fenced-frame-config-instance-struct:~:text=getNestedConfigs()%20method%20steps%20are%3A-,Let%20instance%20be%20this%27s%20relevant%20global%20object%27s%20browsing%20context,If%20instance%20is%20null%2C%20then%20return.,-If%20the%20relevant%20settings

i.e. you check if there is a fenced frame config instance in the browsing context

xyaoinum replied:
Thanks for the pointers. I made an attempt to fix the usage. PTAL again.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +556,204 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+                1. Let |currentTime| be the [=/current wall time=].
+                1. Set |bitDebit|'s [=bit debit/timestamp=] to |currentTime|.
+                1. [=list/Append=] |bitDebit| to |ledger|.
+                1. Set |currentNavigable|'s [=fenced frame config=]'s [=has navigated=] to true.
```
Same response as above.

xyaoinum replied:
I made an attempt to fix the usage. PTAL again.

domfarolino replied:
Can you point me to where this is in the latest PR draft, since this comment is marked as outdated

xyaoinum replied:
If you search for `[=list/Append=] |bitDebit| to |ledger|`, you will find that context.

domfarolino replied:
I don't see anything setting `[=has navigated=]` anymore, did something change?

domfarolino replied:
Resolving per https://github.com/WICG/shared-storage/pull/131#issuecomment-2048069601.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +556,204 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  Between [=beginning navigation=] and [=ending navigation=], a [=user agent=] will perform the [=charge shared storage navigation budget=] algorithm.
+
+  Issue: The [=fenced frame config=] and its boolean <dfn>has navigated</dfn> have not yet been added to the draft [=Fenced Frame=] specification. Some form of them will be added, although their names are subject to bikeshedding. <span class=todo>Fix the names when they are added.</span>
```
It looks like this should be added (also loop the original author @pythagoraskitty), as I don't see a way to check whether the mapping is done (or should I check [mapped-url](https://wicg.github.io/fenced-frame/#fenced-frame-config-mapped-url)'s form? @gtanzer )

xyaoinum replied:
I renamed to "shared storage navigation budget charged" to be more clear, but I still believe it should be added to [=fenced frame config instance=]. @gtanzer to confirm whether this is the case (i.e. IIUC that instance corresponds to `FencedFrameProperties` in the Chrome implementation, and the currently contains a `SharedStorageBudgetMetadata` field)

xyaoinum commented:
@domfarolino: I addressed / responded to your comments (% those that waiting for input from @pythagoraskitty and @gtanzer). PTAL again. Thanks!

domfarolino reviewed (commented):
I left a few one-off comments, so I'll leave a "review comment" to clear my request for a pending review in my dashboard, since we're still waiting on others (and my new comments to be addressed).

xyaoinum commented:
@domfarolino I think all the comments are addresses / has a response. (and one confirmation question for @gtanzer : https://github.com/WICG/shared-storage/pull/131#discussion_r1532579499) PTAL again. Thanks!

xyaoinum commented:
@domfarolino friendly ping. Thanks! (Note also that I just updated this patch to include a "Shared-Storage-Worklet-Allowed: ?1" response header check, which was based on a security review feedback (i.e. relying solely on CORS isn't considered safe enough)).

domfarolino commented:
Can you separate out the header code from this PR to review separately? I just want to avoid piling more things onto this CL which is already massive and difficult to review.

domfarolino commented on spec.bs:
```diff
@@ -201,121 +216,341 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. If the result of running [=determine whether shared storage is allowed by enrollment and user preference=] given |environment| and |workletOrigin| is false:
+        1. If [=this=]'s [=SharedStorageWorklet/is cross-origin worklet=] is false, return a [=promise rejected=] with a {{TypeError}}.
+    1. If |options|["`resolveToConfig`"] is true, [=resolve=] |resultPromise| with |pendingConfig|.
+    1. Othewise, [=resolve=] |resultPromise| to |urn|.
```
```suggestion
    1. Otherwise, [=resolve=] |resultPromise| to |urn|.
```

xyaoinum replied:
Done.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +580,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+        1. Let |url| be the result of running a [=URL parser=] on |urlString|.
+        1. [=Assert=]: |url| is a valid [=/URL=].
+        1. [=map/Set=] |reportingUrlMap|[|eventType|] to |url|.
+    1. <span class=todo>Store |reportingUrlMap| inside a [=fenced frame reporter=] class associated with |fencedFrameConfigStruct|. Both of these still need to be added to the draft [[Fenced-Frame]].</span>
```
Create an issue in the shared storage repo: https://github.com/WICG/shared-storage/issues/144. I suspect this is already doable but I'm not sure. Deferring it to @pythagoraskitty  to figure out, who worked on the reporting part.

xyaoinum replied:
Yes. The location is "Issue(144): Store ..."

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +580,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  ### Navigation Entropy Budget ### {#nav-budget}
+
+  If a user [=user activation|activates=] a [=fenced frame=] whose [=fenced frame config instance=] was generated by {{SharedStorageWorklet/selectURL()}} and thereby initiates a [=top-level traversable=] [=navigate|navigation=], this will reveal to the landing page that its [=/URL=] was selected, which is a leak in [=entropy bits=] of up to logarithm base 2 of the number of input [=/URLs=] for th ...
```
Done. Updated to `... a [=fenced frame=] whose [=Node/node document=]'s [=Document/browsing context=]'s [=browsing context/fenced frame config instance=] ...`

xyaoinum commented:
@domfarolino I addressed the comments, and took out the header part to make this patch smaller. PTAL again. Thanks! Also, responding to your comments below that I'm unable to respond inline:

For comment https://github.com/WICG/shared-storage/pull/131#discussion_r1559612418: I updated it to [=shared storage navigation budget charged=]" which I feel is more clear. 

For https://github.com/WICG/shared-storage/pull/131#issuecomment-2047892683: Yes it's been addressed. The location is: "1. If |options|["`resolveToConfig`"] is true, [=resolve=] |resultPromise| with |pendingConfig|."

domfarolino commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+  [=navigable/traversable navigable=]'s [=traversable navigable/fenced frame config mapping=].
+    1. Let |pendingConfig| be a new [=fenced frame config=].
+    1. Let |urn| be the result of running [=fenced frame config mapping/store a pending config=] on |fencedFrameConfigMapping| with |pendingConfig|.
+    1. If |urn| is failure, then return a [=promise rejected=] with a {{TypeError}}.
```
Can `|url|` be failure? https://wicg.github.io/fenced-frame/#fenced-frame-config-mapping-store-a-pending-config doesn't return a failure in any case, I think.

xyaoinum replied:
It does return a failure.. "2. If the [size](https://infra.spec.whatwg.org/#map-size) of pendingMapping + the [size](https://infra.spec.whatwg.org/#map-size) of mapping’s [finalized config mapping](https://wicg.github.io/fenced-frame/#fenced-frame-config-mapping-finalized-config-mapping) ≥ mapping’s [maximum number of configs](https://wicg.github.io/fenced-frame/#fenced-frame-config-mapping-maximum-number-of-configs), return failure."

domfarolino replied:
Thanks, I missed that.

domfarolino commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. If the result of running [=determine whether shared storage is allowed by enrollment and user preference=] given |environment| and |workletOrigin| is false:
+        1. If [=this=]'s [=SharedStorageWorklet/is cross-origin worklet=] is false, return a [=promise rejected=] with a {{TypeError}}.
+    1. If |options|["`resolveToConfig`"] is true, [=resolve=] |resultPromise| with |pendingConfig|.
+    1. Otherwise, [=resolve=] |resultPromise| to |urn|.
```
```suggestion
    1. Otherwise, [=resolve=] |resultPromise| with |urn|.
```

xyaoinum replied:
Fixed.

domfarolino commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. If the result of running [=determine whether shared storage is allowed by context=] given |environment| and |workletOrigin| is false, return a [=promise rejected=] with a {{TypeError}}.
+    1. If the result of running [=determine whether shared storage is allowed by enrollment and user preference=] given |environment| and |workletOrigin| is false:
+        1. If [=this=]'s [=SharedStorageWorklet/is cross-origin worklet=] is false, return a [=promise rejected=] with a {{TypeError}}.
+    1. If |options|["`resolveToConfig`"] is true, [=resolve=] |resultPromise| with |pendingConfig|.
```
I just want to make sure of something: at the moment we're resolving this with a `|pendingConfig|` that has absolutely nothing filled out (IIUC). Does that match what the implementation does? I seem to recall that the implementation fills _some_ things out before handing it to script, just not the entire thing. Can you or @gtanzer confirm whether this matches what Chromium does / what the WPTs expect?

xyaoinum replied:
Implementation-wise, we indeed return an "empty" struct (with a non-web visible field [urn_uuid_](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/fenced_frame/fenced_frame_config.h?q=%22std::optional%3CKURL%3E%20urn_uuid_%22); and the potentially web-visible fields (sizes) are also not set from shared storage). The WPTs never check the struct field (I assume that's exactly the idea?), they only navigate the fenced frame and assert the landing page.

It'd be good if @gtanzer could further confirm.

gtanzer replied:
Currently there isn't any visibility into the size fields etc. through the IDL config object so it wouldn't be observable anyway. Eventually we probably want to support that; that was part of the idea with configs.

But selectURL doesn't freeze the size, so I don't think there would even be anything to fill out.

domfarolino replied:
OK thanks, I think this can be resolved then?

domfarolino reviewed (commented):
Nothing too major structurally-speaking, which is good. This will probably be the last substantial pass.

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. Let |globalScope| be [=this=]'s [=global scopes=][0].
+    1. Let |workletOrigin| be |globalScope|'s [=global object/realm=]'s \[[HostDefined]] field's [=environment settings object/origin=].
+    1. If the result of running [=Is feature enabled in document for origin?=] on "[=PermissionsPolicy/shared-storage-select-url=]", |document|, and |workletOrigin| returns false, return a [=promise rejected=] with a {{TypeError}}.
+    1. If [=this=]'s [=global scopes=] is [=list/empty=], then return a [=promise rejected=] with a {{TypeError}}.
```
Fixed. Although, I believe it should be `This can happen if either {{WindowSharedStorage/selectURL()}} or {{SharedStorageWorklet/selectURL()}} is called before {{addModule()}}`. This isn't applicable to createWorklet(), because createWorklet() will create the global scope and the (new) worklet object at the same time, and thus it's guaranteed that selectURL() is called after that.

domfarolino replied:
Ah, you're saying selectURL() _cannot_ be called after createWorklet() then, right?

xyaoinum replied:
selectURL() _can_ be called after createWorklet(). Not sure which selectURL() you are referring to, but:
- `sharedStorage.selectURL()` will defer to `sharedStorage.worklet.selectURL()`
- For an explicitly created SharedStorageWorklet object (via `let worklet = await sharedStorage.createWorklet()`, the new worklet object should also be able to call `worklet.selectURL()` after creation.

domfarolino replied:
Oops I'm sorry. I meant "_before_" in my previous comment.

xyaoinum replied:
It'd be an ambiguous statement. `createWorklet()` operates on `window.sharedStorage`, and its _result_ is a `SharedStorageWorklet` object, which can further invoke `worklet.selectURL()`. But if you are talking about `sharedStorage.createWorklet()` v.s `sharedStorage.selectURL()`, they are _unrelated_, because they create/operate on different worklet object.

xyaoinum commented on spec.bs:
```diff
@@ -326,18 +545,16 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
     1. Let |operationMap| be this {{SharedStorageWorkletGlobalScope}}'s [=SharedStorageWorkletGlobalScope/operation map=].
```
Yes. Also added "Issue(151)" inline in the spec.

xyaoinum commented on spec.bs:
```diff
@@ -326,18 +545,16 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
     1. Let |operationMap| be this {{SharedStorageWorkletGlobalScope}}'s [=SharedStorageWorkletGlobalScope/operation map=].
     1. If |operationMap| [=map/contains=] an [=map/entry=] with [=map/key=] |name|, throw a {{TypeError}}.
     1. If |operationCtor| is missing, throw a {{TypeError}}.
```
Added "Issue(151)" inline in the spec.

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+
     1. If the result of running [=obtaining a site|obtain a site=] with |origin| is not [=enrolled=], then return false.
-    1. If the result of running [=check if user preference setting allows access to shared storage=] from |environment| is false, then return false.
+    1. If the result of running [=check if user preference setting allows access to shared storage=] given |environment| and |origin| is false, then return false.
```
Done.

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+
+    - For each method under [[#window-setter]], |environment| is the current context, and |origin| is |environment|'s [=environment settings object/origin=].
+    - For creating a worklet, |environment| is the [=environment settings object=] associated with the {{Window}} that created the worklet, and |origin| is the module scirpt url's [=url/origin=].
+    - For initiating (from a {{Window}}) and for running (from {{SharedStorageWorkletGlobalScope}}) operations on a worklet, |environment| is the [=environment settings object=] associated with the {{Window}} that created the worklet, and |origin| is the worklet's [=global scopes=][0]'s [=global object/realm=]'s \[[HostDefined]] field's [=environment settings object/origin=].
```
Ack that this was indeed confusing. I updated it to "For running operations on a worklet (from a {{Window}}), and for each method under [[#worklet-setter]] (from {{SharedStorageWorkletGlobalScope}}), ..."

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+            Note: This indicates that |operationCtor|'s run() method encounters an error, where |operationCtor| is the parameter in {{SharedStorageWorkletGlobalScope/register()}}.
+
+        1. Otherwise:
+            1. Set |index| to |operationResult|.
```
Done. Un-nested.

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+        1. Otherwise:
+            1. Set |index| to |operationResult|.
+            1. If [=an exception was thrown=]:
+                1. [=Queue a global task=] on the [=DOM manipulation task source=], given |window|, to [=reject=] |promise| with a {{TypeError}}.
```
Done.

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+                Note: The result index is a non-integer value. This violates the selectURL() protocol, and we don't know which url should be selected.
+
+            1. Otherwise, if |index| is greater than |urlList|'s [=list/size=]:
+                1. [=Queue a global task=] on the [=DOM manipulation task source=], given |window|, to [=reject=] |promise| with a {{TypeError}}.
```
Done.

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. Let |document| be |context|'s [=active document=].
+    1. [=Assert=]: [=this=]'s [=global scopes=]'s [=list/size=] is 1.
+    1. Let |globalScope| be [=this=]'s [=global scopes=][0].
+    1. Let |workletOrigin| be |globalScope|'s [=global object/realm=]'s \[[HostDefined]] field's [=environment settings object/origin=].
```
Done.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  If a user [=user activation|activates=] a [=fenced frame=] whose [=Node/node document=]'s [=Document/browsing context=]'s [=browsing context/fenced frame config instance=] was generated by {{SharedStorageWorklet/selectURL()}} and thereby initiates a [=top-level traversable=] [=navigate|navigation=], this will reveal to the landing page that its [=/URL=] was selected, which is a leak in [=entrop ...
+
+  A <dfn>calling site</dfn> for {{SharedStorageWorklet/selectURL()}} is a [=site=].
```
Filed https://github.com/WICG/shared-storage/issues/147 and added the Issue(147) in the spec as well.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  When a leak occurs, its value in [=entropy bits=] is calculated and stored for that [=calling site=], along with the current time as a [=bit debit/timestamp=], together as a [=bit debit=] in the [=shared storage navigation budget table=].
+
+  A [=calling site=]'s <dfn for="calling site">remaining navigation budget</dfn> is the [=navigation entropy allowance=] minus any [=bit debits=] whose [=bit debit/timestamps=] are within the current [=navigation budget epoch=].
```
Filed https://github.com/WICG/shared-storage/issues/147 and added the Issue(147) in the spec as well.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  A [=calling site=]'s <dfn for="calling site">remaining navigation budget</dfn> is the [=navigation entropy allowance=] minus any [=bit debits=] whose [=bit debit/timestamps=] are within the current [=navigation budget epoch=].
+
+  {{SharedStorageWorklet/selectURL()}}'s argument "`urls`" is its <dfn for=selectURL>input URL list</dfn>.
```
Filed https://github.com/WICG/shared-storage/issues/147 and added the Issue(147) in the spec as well.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  When a [=calling site=] has insufficient [=calling site/remaining navigation budget=], {{SharedStorageWorklet/selectURL()}} will return a {{SharedStorageResponse}} (i.e. either a {{FencedFrameConfig}} or a [=urn uuid=]) for the {{SharedStorageUrlWithMetadata/url}} in the {{SharedStorageUrlWithMetadata}} at the [=default index=] in its [=selectURL/input URL list=].
+
+  The <dfn>default index</dfn> for a call to {{SharedStorageWorklet/selectURL()}} is [=implementation-defined=] in such a way that it is independent from the result of the registered operation class's "`run`" method.
```
Filed https://github.com/WICG/shared-storage/issues/147 and added the Issue(147) in the spec as well.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+    1. Let |currentNavigable| be |sourceDocument|'s [=node navigable=].
+    1. While |currentNavigable| is not null:
+        1. Let |site| be the result of running [=obtain a site=] with |currentNavigable|'s [=active document=]'s [=document/origin=].
+        1. Let |instance| be |currentNavigable|'s [=Node/node document=]'s [=Document/browsing context=]'s [=browsing context/fenced frame config instance=].
```
Done. Switched to "Let |instance| be |currentNavigable|'s [=source snapshot params/initiator fenced frame config instance=]."

domfarolino replied:
Great!

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+  Issue: The boolean <dfn>shared storage reporting budget charged</dfn> have not yet been added to [=fenced frame config instance=] in the draft [[Fenced-Frame]] specification. Some form of them will be added, although their names are subject to bikeshedding. <span class=todo>Fix the names when they are added.</span>
+
+  <div algorithm>
+    To <dfn>charge shared storage reporting budget</dfn> given a {{Document}} |sourceDocument|, run the following steps:
```
Added and referenced Issue(150).

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+
+  The algorithm [=fetch a worklet script graph=] calls into the <a href="https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-worklet/module-worker-script-graph">fetch a worklet/module worker script graph</a> algorithm, which takes in an algorithm parameter |processCustomFetchResponse|. The definition of that |processCustomFetchResponse| parameter will need to include the following step ...
+
+    5. If <var ignore=''>fetchClient</var>'s [=environment settings object/realm execution context=]'s Realm component's [=global object=] is {{SharedStorageWorkletGlobalScope}}:
```
Yes, updated to "<var ignore=''>fetchClient</var>'s [=environment settings object/global object=]"

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+        1. If |urlWithMetadata| has field "`reportingMetadata`":
+            1. Let |reportingMetadata| be |urlWithMetadata|["`reportingMetadata`"].
+            1. If the result of running [=validate reporting metadata=]  with |reportingMetadata| is false, [=reject=] |resultPromise| with a {{TypeError}} and abort these steps.
+    1. Let |fencedFrameConfigMapping| be |globalScope|'s [=associated Document=]'s [=node navigable=]'s
```
Yes. Good catch. Updated to "|window|'s [=associated document=]'s [=node navigable=]'s ...."

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+        1. If |urlWithMetadata| has field "`reportingMetadata`":
+            1. Let |reportingMetadata| be |urlWithMetadata|["`reportingMetadata`"].
+            1. If the result of running [=validate reporting metadata=]  with |reportingMetadata| is false, [=reject=] |resultPromise| with a {{TypeError}} and abort these steps.
+    1. Let |fencedFrameConfigMapping| be |globalScope|'s [=associated Document=]'s [=node navigable=]'s
```
Good catch. Fixed. I also took a pass and fixed one other place.

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+    This rationale also applies to the handling for user preferences error for {{SharedStorageWorklet/selectURL()}} and {{SharedStorageWorklet/run()}}.
+    </div>
+
+    <span class=todo>Specify that redirects are disallowed when fetching the module script for {{SharedStorageWorklet}}.</span>
```
Yes. This is resolved. Deleted.

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+
+  The {{SharedStorageWorklet}}'s [=worklet destination type=] is "sharedstorageworklet".
+
+  <span class=todo>Add "sharedstorageworklet" to the possible strings that a request [=request/destination=] can have.</span>
```
Linked.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
 
+  ## {{SharedStorageUrlWithMetadata}} and Reporting ## {#reporting}
+
+  {{SharedStorageUrlWithMetadata}} [=dictionaries=] contain a [=string=] representing a [=/URL=] and, optionally, a {{reportingMetadata}} {{/object}}.
```
Ack. Removed.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+    };
+  </xmp>
+
+  If a {{SharedStorageUrlWithMetadata}} [=dictionary=] contains a non-[=map/empty=] {{SharedStorageUrlWithMetadata/reportingMetadata}} {{/object}} in the form of a [=dictionary=] whose [=map/keys=] are {{FenceEvent/eventType}}s and whose [=map/values=] are [=strings=] that parse to valid [=/URLs=], then these {{FenceEvent/eventType}}-[=/URL=] pairs will be [=register reporting metadata|registered ...
```
I just barely updated it to _"...whose [=map/keys=] are {{FenceEvent}}'s {{FenceEvent/eventType}}s and..."_. I think none of the fields are directly related to a {{FencedEvent}} when provided by the user. It will be used to construct a FencedEvent at a later stage. This is nevertheless underspecified, and under section "To <dfn>register reporting metadata</dfn>", I've linked to issue https://github.com/WICG/shared-storage/issues/144.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  A <dfn>calling site</dfn> for {{SharedStorageWorklet/selectURL()}} is a [=site=].
+
+  A <dfn>navigation entropy allowance</dfn> is a maximum allowance of [=entropy bits=] that are permitted to leak via [=fenced frames=] initiating [=top-frame=] [=navigate|navigations=] during a given [=navigation budget epoch=] for a given calling [=calling site=]. This [=navigation entropy allowance|allowance=] is defined by the [=user agent=] and is [=calling site=]-agnostic.
```
Fixed. switched to [=top-level traversable=] everywhere.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+  <div algorithm>
+    To <dfn>determine remaining navigation budget</dfn>, given an [=environment settings object=] |environment| and a [=calling site=] |site|, run the following steps:
+
+    1. If |site| is an [=opaque origin=], return undefined.
```
Yes. I changed to "[=Assert=]: |site| is not an [=opaque origin=]" here, and removed the outside asserts.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+    1. [=map/iterate|For each=] [=list/item=] |bitDebit| in |ledger|, do the following steps:
+        1. Let |debit| be |bitDebit|'s [=bit debit/bits=].
+        1. If the result of running [=check whether a bit debit is expired=] with |environment| and |bitDebit| is false, then increment |debitSum| by |debit|.
+    1. Return |maxBits| minus |debitSum|.
```
Done.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  Between [=beginning navigation=] and [=ending navigation=], a [=user agent=] will perform the [=charge shared storage navigation budget=] algorithm.
+
+  Issue: The boolean <dfn>shared storage navigation budget charged</dfn> have not yet been added to [=fenced frame config instance=] in the draft [[Fenced-Frame]] specification. Some form of them will be added, although their names are subject to bikeshedding. <span class=todo>Fix the names when they are added.</span>
```
Added Issue(149)

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+        1. If |instance| is not null and |site| is not an [=opaque origin=], perform the following steps:
+            1. Let |pendingBits| be |instance|'s [=pending shared storage budget debit=].
+            1. If |pendingBits| is greater than 0 and if |instance|'s [=shared storage navigation budget charged=] is false, run the following steps:
+                1. Let |ledger| be [=user agent=]'s [=shared storage navigation budget table=][|site|].
```
Done.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+    1. While |currentNavigable| is not null:
+        1. Let |instance| be |currentNavigable|'s [=Node/node document=]'s [=Document/browsing context=]'s [=browsing context/fenced frame config instance=].
+        1. If |instance| is not null:
+            1. Let |pendingBits| be |instance|'s [=pending shared storage budget debit=].
```
Done.

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+  </div>
+
+
+  Issue: The boolean <dfn>shared storage reporting budget charged</dfn> have not yet been added to [=fenced frame config instance=] in the draft [[Fenced-Frame]] specification. Some form of them will be added, although their names are subject to bikeshedding. <span class=todo>Fix the names when they are added.</span>
```
Added Issue(149)

xyaoinum commented:
@domfarolino I've addressed / responded to your comments. PTAL again. Thanks!
@gtanzer Please confirm if my understanding in https://github.com/WICG/shared-storage/pull/131#discussion_r1563622906 is correct. Thanks!

xyaoinum commented on spec.bs:
```diff
@@ -347,6 +564,209 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
+
+  A [=bit debit=] will need to be [=charge shared storage navigation budget|charged=] to the [=shared storage navigation budget table=] for each [=top-frame=] [=navigate|navigation=] initiated by a [=fenced frame=] whose [=Node/node document=]'s [=Document/browsing context=]'s [=browsing context/fenced frame config instance=] was generated via {{SharedStorageWorklet/selectURL()}}, as this can lea ...
+
+  <span class=todo>Move the definition of [=pending shared storage budget debit=] to [=fenced frame config instance=] in the draft [[Fenced-Frame]] specification.</span>
```
Added Issue(148)

xyaoinum commented on spec.bs:
```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+    5. If <var ignore=''>fetchClient</var>'s [=environment settings object/realm execution context=]'s Realm component's [=global object=] is {{SharedStorageWorkletGlobalScope}}:
+        1. Set |request|'s [=request/redirect mode=] to "<code>error</code>".
+
+            Note: For shared storage, redirects are disallowed for the module script request. With this restriction, it's possible to define and to use the algorithm that gets the |realm|'s \[[HostDefined]] field's [=environment settings object/origin=] (as described in [[#set-up-a-worklet-environment-settings-object-monkey-patch]]) as soon as the {{SharedStorageWorkletGlobalScope}} is created, a ...
```
Done.

domfarolino reviewed (approved):
OK I think we've addressed the meat of worklet internals with the previous reviews here so far, so I'm happy moving forward with this due to the urgency of this work, and thank you very much for sticking with it as well as filing the follow-up issues for the small remaining things that can be addressed separately.

Please send PRs _soon_ to fix the following:
 - https://github.com/WICG/shared-storage/issues/151
 - https://github.com/WICG/shared-storage/issues/150
 - https://github.com/WICG/shared-storage/issues/149
 - https://github.com/WICG/shared-storage/issues/148
 - https://github.com/WICG/shared-storage/issues/147
 - https://github.com/WICG/shared-storage/issues/146
 - https://github.com/WICG/shared-storage/issues/145
 - https://github.com/WICG/shared-storage/issues/135
 - https://github.com/WICG/shared-storage/issues/134

More smaller PRs would be better than larger PRs combining the issues above.

xyaoinum commented on spec.bs:
```diff
@@ -1165,15 +1333,16 @@ On the other hand, methods for getting data from the [=shared storage database=]
     1. Let |realm| be the [=current realm=].
+    1. If the result of running [=determine whether shared storage is allowed by context=] given |environment| and |realm|'s <span class=todo>HostDefined</span> field's [=environment settings object/origin=] is false, return a [=promise rejected=] with a {{TypeError}}.
+    1. If the result of running [=determine whether shared storage is allowed by enrollment and user preference=] given |environment| and |realm|'s <span class=todo>HostDefined</span> field's [=environment settings object/origin=] is false, return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |site| be the result of running [=obtain a site=] with |realm|'s <span class=todo>HostDefined</span> field's [=environment settings object/origin=].
```
I think this was already done (did you look at a previous version of the file?)

domfarolino replied:
I just saw that the comment was not resolved.

xyaoinum replied:
Got it. This should have been resolved.

xyaoinum commented:
Thanks for the thorough review @domfarolino . (For the last few comments, I think you might looked at an outdated version of the patch? The comments in question should have already been addressed.)

jkarlin commented:
Applies to issue #2
